src/base58.cpp

```diff
@@ -313,39 +313,35 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
 }
 } // namespace
 
-void CBitcoinSecret::SetKey(const CKey& vchSecret)
+CKey DecodeSecret(const std::string& str)
 {
-    assert(vchSecret.IsValid());
-    SetData(Params().Base58Prefix(CChainParams::SECRET_KEY), vchSecret.begin(), vchSecret.size());
-    if (vchSecret.IsCompressed())
-        vchData.push_back(1);
+    CKey key;
+    std::vector<unsigned char> data;
+    if (DecodeBase58Check(str, data)) {
+        const std::vector<unsigned char>& privkey_prefix = Params().Base58Prefix(CChainParams::SECRET_KEY);
+        if ((data.size() == 32 + privkey_prefix.size() || (data.size() == 33 + privkey_prefix.size() && data.back() == 1)) &&
+            std::equal(privkey_prefix.begin(), privkey_prefix.end(), data.begin())) {
+            bool compressed = data.size() == 33 + privkey_prefix.size();
+            key.Set(data.begin() + privkey_prefix.size(), data.begin() + privkey_prefix.size() + 32, compressed);
+        }
+    }
+    memory_cleanse(data.data(), data.size());
```

Comment:
Important: IMO `DecodeBase58Check`should *cleanse* `data`before returning `false`.

Reply:
Extra: In that case this function could be:
```cpp
    CKey key;
    std::vector<unsigned char> data;
    if (!DecodeBase58Check(str, data)) return key;

    const auto& prefix = Params().Base58Prefix(CChainParams::SECRET_KEY);
    int size = data.size() - prefix.size();
    if (size == 32 || (size == 33 && data.back() == 1)) {
        if (std::equal(prefix.begin(), prefix.end(), data.begin())) {
            ...
        }
    }

    memory_cleanse(data.data(), data.size());
    return key;
```

Reply:
Side note: Another option is to create a PR to use custom allocator for all containers that are `memory_cleanse`.